### PR TITLE
Detect subpaths for dirs

### PR DESCRIPTION
**Pull request by mdales**
I was mocking up LIFE, and in that there's a dir created from which we select a single file later, and I want to have that flow treated as the dir is an input, with a labelled edge that tells of the specific  file used.

aoh

### CODE REVIEW

File: src/lib/ast.ml [hunk 1]
```diff
@@ -1,31 +1,80 @@
...
+  type t = { id : int; path : string; subpath : string option; wildcard : bool }
+  [@@deriving sexp]
+
+  let pp ppf t = Sexplib.Sexp.pp_hum ppf (sexp_of_t t)
+
+  let v ?(subpath = None) id path =
+    let wildcard =
+      match subpath with
+      | None -> false
+      | Some p ->
```
Comment (patricoferris) [inline src/lib/ast.ml:14]: Could be simplified I think to

```ocaml
| Some p -> Char.equal p.[String.length p - 1] '*'
```

File: src/lib/ast.ml [hunk 2]
```diff
@@ -1,31 +1,80 @@
...
+          let last_char =
+            String.Sub.to_string (String.sub ~start:(String.length p - 1) p)
+          in
+          last_char = "*"
+    in
+    if wildcard then
+      {
+        id;
+        path = String.Sub.to_string (String.sub ~stop:(String.length path) path);
```
Comment (patricoferris) [inline src/lib/ast.ml:23]: Doesn't this produce the same string? Was the intent to drop the last character of the string?

File: src/lib/ast.ml [hunk 3]
```diff
@@ -35,10 +84,38 @@ type t = CommandGroup.t list
 
 let to_list cg = cg
 
+let find_matching_datafile datafile_map path =
+  match List.assoc_opt path datafile_map with
+  | Some p -> Some p
+  | None ->
+      (* No full match, but can we find a prefix dir *)
+      List.fold_left
+        (fun acc i ->
```
Comment (patricoferris) [inline src/lib/ast.ml:93]: More idiomatic (read please ignore if you want ^^) to pull out the `(ipath, df)` here

```suggestion
        (fun acc (ipath, df) ->
```

File: src/lib/ast.mli [hunk 4]
```diff
@@ -9,25 +9,36 @@ module DataFile : sig
 
   type t
 
-  val create : int -> string -> t
+  val v : ?subpath:string option -> int -> string -> t
```
Comment (patricoferris) [inline src/lib/ast.mli:12]: Optional arguments will default to `None`.

```suggestion
  val v : ?subpath:string -> int -> string -> t
```

File: src/lib/ast.ml [hunk 5]
```diff
@@ -1,31 +1,80 @@
...
+open Sexplib.Conv
+
 module DataFile = struct
-  type t = { id : int; path : string }
+  type t = { id : int; path : string; subpath : string option; wildcard : bool }
+  [@@deriving sexp]
+
+  let pp ppf t = Sexplib.Sexp.pp_hum ppf (sexp_of_t t)
+
+  let v ?(subpath = None) id path =
```
Comment (patricoferris) [inline src/lib/ast.ml:10]: With the update I mentioned in the `ast.mli` file this becomes:

```suggestion
  let v ?subpath id path =
```

File: src/test/test.ml [hunk 6]
```diff
@@ -103,6 +103,80 @@ module CommandParsing = struct
...
+      "No subpath" None
+      (Shark.Ast.DataFile.subpath test);
+    Alcotest.(check bool)
+      "Isn't wildcard" false
+      (Shark.Ast.DataFile.is_wildcard test);
+    Alcotest.(check bool) "Is dir" true (Shark.Ast.DataFile.is_dir test)
+
+  let test_basic_dir_canonical_with_wildcard () =
+    let testcase = "/data/test/" in
+    let test = Shark.Ast.DataFile.v ~subpath:(Some "*") 42 testcase in
```
Comment (patricoferris) [inline src/test/test.ml:152]: With the updates to `?subpath` this becomes

```suggestion
    let test = Shark.Ast.DataFile.v ~subpath:"*" 42 testcase in
```

**Review by patricoferris — COMMENTED** [review]
Thanks! Looking good, all this file path wrangling does make me think perhaps using something like [fpath](https://ocaml.org/p/fpath/latest) might be useful.

File: src/lib/ast.ml (hunk 3 above)
Comment (mdales) [reply src/lib/ast.ml:93]: I thought I could do that, had tried it, and failed. I suspect that what happened was I had an error somewhere else, that I then accidentally cleared up when I moved to this form, and so I got a false positive on this pattern being the right way. Thanks!

**Review by patricoferris — APPROVED** [review]
This looks good to me, I didn't do a thorough check of the logic but the OCaml is good :))

**Pull request merged**